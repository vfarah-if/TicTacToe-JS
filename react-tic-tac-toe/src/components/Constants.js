export const TOP_LEFT = 0;
export const TOP_MIDDLE = 1;
export const TOP_RIGHT = 2;
export const MIDDLE_LEFT = 3;
export const MIDDLE = 4;
export const MIDDLE_RIGHT = 5;
export const BOTTOM_LEFT = 6;
export const BOTTOM_MIDDLE = 7;
export const BOTTOM_RIGHT = 8;
export const LINES = [
    // Horizontal Lines
    [TOP_LEFT, TOP_MIDDLE, TOP_RIGHT],
    [MIDDLE_LEFT, MIDDLE, MIDDLE_RIGHT],
    [BOTTOM_LEFT, BOTTOM_MIDDLE, BOTTOM_RIGHT],
    // Vertical Lines
    [TOP_LEFT, MIDDLE_LEFT, BOTTOM_LEFT],
    [TOP_MIDDLE, MIDDLE, BOTTOM_MIDDLE],
    [BOTTOM_RIGHT, MIDDLE_RIGHT, TOP_RIGHT],
    // Diagonal Lines
    [TOP_LEFT, MIDDLE, BOTTOM_RIGHT],
    [TOP_RIGHT, MIDDLE, BOTTOM_LEFT]
];