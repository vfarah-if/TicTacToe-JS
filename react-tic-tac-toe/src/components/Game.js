import React, { Component } from 'react';
import Board from './Board';
import {
    LINES
} from './Constants';

export default class Game extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isPlayerOneActive: true,
            playerOne: 'X',
            playerTwo: '0',
            history: [
                { squares: Array(9).fill(null) }
            ],
            winner: null
        };
    }

    handleClick(i) {
        if (!this.state.winner) {
            const { history } = this.state;
            const current = history[history.length -1];
            const squares = current.squares.slice();
            if (this.isEmpty(squares[i])) {
                squares[i] = this.state.isPlayerOneActive ? this.state.playerOne : this.state.playerTwo;
                const winner = this.calculateWinner(squares);
                this.updateState(history, squares, !this.state.isPlayerOneActive, winner);
            }
        }
    }

    isEmpty(square) {
        return square === null;
    }

    updateState(history, squares, isPlayerOneActive, winner) {
        this.setState({
            history: history.concat([{
                squares: squares
              }]),
            isPlayerOneActive: isPlayerOneActive,
            winner: winner
        });
    }

    calculateWinner(squares) {
        for (let index = 0; index < LINES.length; index++) {
            const [a, b, c] = LINES[index];
            if (this.lineIsUniform(squares, a, b, c)) {
                return squares[a];
            }
        }

        return null;
    }

    lineIsUniform(squares, a, b, c) {
        return squares[a] && squares[a] === squares[b] && squares[a] === squares[c];
    }

    render() {
        const { isPlayerOneActive, playerOne, playerTwo, winner, history } = this.state;
        const current = history[history.length - 1];
        let status = `Next player: ${isPlayerOneActive ? playerOne : playerTwo}`;
        if (winner) {
            status = `Winner: ${winner}`;
        }

        return (
            <div className="game">
                <div className="game-board">
                    <Board squares={current.squares} onClick={(i) => this.handleClick(i)} />
                </div>
                <div className="game-info">
                    <div className="status">{status}</div>
                </div>
            </div>

        );
    }
}
