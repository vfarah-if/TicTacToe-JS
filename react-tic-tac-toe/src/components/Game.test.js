import React from 'react';
import { shallow } from 'enzyme';

import Game from './Game';
import { TOP_LEFT } from './Constants';
// import the test scenarios from a file to generate repetitive tests, see test at bottom for more
import testScenarios from './Game.test.scenarios.json';

function getBoardElement(wrapper) {
    return wrapper.find('game.game-board>Board');
}

function getStatus(wrapper) {
    return wrapper.find('div.status');
}

describe('Game', () => {
    it('should create a board', () => {
        const wrapper = shallow(<Game />);
        const boardElement = getBoardElement(wrapper);
        expect(boardElement).toBeTruthy();
    });

    it('should set X or player one as the first player', () => {
        const wrapper = shallow(<Game />)
        const status = getStatus(wrapper);
        expect(status).toBeTruthy();
        expect(status.text()).toBe('Next player: X')
    });

    it('should toggle active players', () => {
        const wrapper = shallow(<Game />)
        let status = getStatus(wrapper);
        expect(status.text()).toBe('Next player: X')
        wrapper.instance().handleClick(TOP_LEFT);
        status = getStatus(wrapper);
        expect(status.text()).toBe('Next player: 0')
    });

    it('should prevent the same position being edited more than once', () => {
        const wrapper = shallow(<Game />)
        expect(getStatus(wrapper).text()).toBe('Next player: X')
        wrapper.instance().handleClick(TOP_LEFT);
        wrapper.update();
        expect(getStatus(wrapper).text()).toBe('Next player: 0')

        wrapper.instance().handleClick(TOP_LEFT);
        wrapper.update();
        
        expect(getStatus(wrapper).text()).toBe('Next player: 0')
    });

    // Remarks : Utilised json-loader to load scenarios from a json file
    testScenarios.forEach((theory: any) => {
        it(theory.scenario, () => {
            const wrapper = shallow(<Game />)
            console.log('Scenario', theory.data, theory.expectedWinner);
            theory.data.forEach(cell => {
                wrapper.instance().handleClick(cell);
            });
            console.log('Example', theory.example);            
            expect(getStatus(wrapper).text()).toBe(theory.expectedWinner)
        });
    });
});