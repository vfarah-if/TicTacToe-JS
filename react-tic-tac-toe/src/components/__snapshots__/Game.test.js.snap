// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should create a board and set x as the first player 1`] = `
Array [
  <div
    className="game-board"
  >
    <Board
      onClick={[Function]}
      squares={
        Array [
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
        ]
      }
    />
  </div>,
  <div
    className="game-info"
  >
    <div
      className="status"
    >
      Next player: X
    </div>
  </div>,
]
`;

exports[`Game should toggle active players with create 1`] = `
<div
  className="game"
>
  <div
    className="game-board"
  >
    <div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        >
          X
        </button>
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
  <div
    className="game-info"
  >
    <div
      className="status"
    >
      Next player: 0
    </div>
  </div>
</div>
`;
